quickwit/quickwit-search/src/top_k_collector.rs

```diff
@@ -0,0 +1,860 @@
+                Option<u64>,
+                Option<u64>,
+                false,
+            >::new(
+                split_id, score_extractor, leaf_max_hits, segment_ord
+            ))
+        }
+    }
+}
+
+/// Fast Top K Computation
+///
+/// The buffer is truncated to the top_n elements when it reaches the capacity of the Vec.
+/// That means capacity has special meaning and should be carried over when cloning or serializing.
+///
+/// For TopK == 0, it will be relative expensive.
+pub struct TopKComputer<D> {
+    /// Reverses sort order to get top-semantics instead of bottom-semantics
+    buffer: Vec<Reverse<D>>,
+    top_n: usize,
+    pub(crate) threshold: Option<D>,
+}
+
+// Custom clone to keep capacity
+impl<D: Clone> Clone for TopKComputer<D> {
+    fn clone(&self) -> Self {
+        let mut buffer_clone = Vec::with_capacity(self.buffer.capacity());
+        buffer_clone.extend(self.buffer.iter().cloned());
+
+        TopKComputer {
+            buffer: buffer_clone,
+            top_n: self.top_n,
+            threshold: self.threshold.clone(),
+        }
+    }
+}
+
+impl<D> TopKComputer<D>
+where D: Ord + Clone + Debug
+{
+    /// Create a new `TopKComputer`.
+    /// Internally it will allocate a buffer of size `5 * top_n`.
+    pub fn new(top_n: usize) -> Self {
+        // Vec cap can't be 0, since it would panic in push
+        let vec_cap = top_n.max(1) * 10;
+        TopKComputer {
+            buffer: Vec::with_capacity(vec_cap),
+            top_n,
+            threshold: None,
+        }
+    }
+
+    /// Push a new document to the top n.
+    /// If the document is below the current threshold, it will be ignored.
+    #[inline]
+    pub fn push(&mut self, doc: D) {
+        if let Some(last_median) = self.threshold.clone() {
+            if doc < last_median {
+                return;
+            }
+        }
+        if self.buffer.len() == self.buffer.capacity() {
+            let median = self.truncate_top_n();
+            self.threshold = Some(median);
+        }
+
+        // This is faster since it avoids the buffer resizing to be inlined from vec.push()
+        // (this is in the hot path)
+        // TODO: Replace with `push_within_capacity` when it's stabilized
+        let uninit = self.buffer.spare_capacity_mut();
+        // This cannot panic, because we truncate_median will at least remove one element, since
+        // the min capacity is larger than 2.
+        uninit[0].write(Reverse(doc));
+        // This is safe because it would panic in the line above
+        unsafe {
+            self.buffer.set_len(self.buffer.len() + 1);
+        }
+    }
+
+    #[inline(never)]
+    fn truncate_top_n(&mut self) -> D {
+        // Use select_nth_unstable to find the top nth score
+        let (_, median_el, _) = self.buffer.select_nth_unstable(self.top_n);
+
+        let median_score = median_el.clone();
+        // Remove all elements below the top_n
+        self.buffer.truncate(self.top_n);
```

Comment:
do you know if truncate_top_n or `self.buffer.truncate` takes a lot of perf?

Reply:
`self.buffer.truncate` is just setting the len and therefore very cheap, but `self.buffer.select_nth_unstable(self.top_n)` takes the majority of time on some queries, especially if the threshold is not effective